

> Classification


---

In my model for the dataset titeled "Diabetes", i will predict whether a person has diabetes or not based on many factors.



#Importing Librairies

In [1]:
import numpy as np
import pandas as pd


#Importing dataset

In [2]:
df=pd.read_csv("Diabetes.csv")
df

Age  Gender  BMI Blood Pressure  FBS  HbA1c Family History of Diabetes  \
0     45    Male   25         Normal  100    5.7                         No   
1     55  Female   30           High  120    6.4                        Yes   
2     65    Male   35           High  140    7.1                        Yes   
3     75  Female   40           High  160    7.8                        Yes   
4     40    Male   20         Normal   80    5.0                         No   
..   ...     ...  ...            ...  ...    ...                        ...   
123   17  Female   15         Normal  100    5.7                         No   
124   22    Male   19         Normal  120    6.4                         No   
125   27  Female   24           High  140    7.1                         No   
126   32    Male   29           High  160    7.8                         No   
127   37  Female   34           High  180    8.5                         No   

    Smoking     Diet Exercise Diagnosis  
0        No  Healthy  Regular        No  
1       Yes     Poor       No       Yes  
2       Yes     Poor       No       Yes  
3       Yes     Poor       No       Yes  
4        No  Healthy  Regular        No  
..      ...      ...      ...       ...  
123     Yes     Poor       No       Yes  
124     Yes     Poor       No       Yes  
125     Yes     Poor       No       Yes  
126     Yes     Poor       No       Yes  
127     Yes     Poor       No       Yes  

[128 rows x 11 columns]

In [45]:
x = df.iloc[:, :-1]
y = df.iloc[:, -1]

In [46]:
x.head()

Age  Gender  BMI Blood Pressure  FBS  HbA1c Family History of Diabetes  \
0   45    Male   25         Normal  100    5.7                         No   
1   55  Female   30           High  120    6.4                        Yes   
2   65    Male   35           High  140    7.1                        Yes   
3   75  Female   40           High  160    7.8                        Yes   
4   40    Male   20         Normal   80    5.0                         No   

  Smoking     Diet Exercise  
0      No  Healthy  Regular  
1     Yes     Poor       No  
2     Yes     Poor       No  
3     Yes     Poor       No  
4      No  Healthy  Regular

In [47]:
x.shape

(128, 10)

In [48]:
y

0       No
1      Yes
2      Yes
3      Yes
4       No
      ... 
123    Yes
124    Yes
125    Yes
126    Yes
127    Yes
Name: Diagnosis, Length: 128, dtype: object

In [49]:
y.shape

(128,)

#Data understanding

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128 entries, 0 to 127
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         128 non-null    int64  
 1   Gender                      128 non-null    object 
 2   BMI                         128 non-null    int64  
 3   Blood Pressure              128 non-null    object 
 4   FBS                         128 non-null    int64  
 5   HbA1c                       128 non-null    float64
 6   Family History of Diabetes  128 non-null    object 
 7   Smoking                     128 non-null    object 
 8   Diet                        128 non-null    object 
 9   Exercise                    128 non-null    object 
 10  Diagnosis                   128 non-null    object 
dtypes: float64(1), int64(3), object(7)
memory usage: 11.1+ KB


In [51]:
df.shape

(128, 11)

In [52]:
df.describe()

Age         BMI         FBS       HbA1c
count  128.000000  128.000000  128.000000  128.000000
mean    42.031250   35.359375  162.500000    7.887500
std     16.783915   14.981739   61.323975    2.146339
min     12.000000   10.000000   80.000000    5.000000
25%     28.000000   24.000000  120.000000    6.400000
50%     40.000000   34.000000  160.000000    7.800000
75%     55.000000   45.500000  205.000000    9.375000
max     75.000000   67.000000  280.000000   12.000000

#Data preprocessing

##Missing values

In [53]:
df.isnull().sum()

Age                           0
Gender                        0
BMI                           0
Blood Pressure                0
FBS                           0
HbA1c                         0
Family History of Diabetes    0
Smoking                       0
Diet                          0
Exercise                      0
Diagnosis                     0
dtype: int64

In [54]:
df

Age  Gender  BMI Blood Pressure  FBS  HbA1c Family History of Diabetes  \
0     45    Male   25         Normal  100    5.7                         No   
1     55  Female   30           High  120    6.4                        Yes   
2     65    Male   35           High  140    7.1                        Yes   
3     75  Female   40           High  160    7.8                        Yes   
4     40    Male   20         Normal   80    5.0                         No   
..   ...     ...  ...            ...  ...    ...                        ...   
123   17  Female   15         Normal  100    5.7                         No   
124   22    Male   19         Normal  120    6.4                         No   
125   27  Female   24           High  140    7.1                         No   
126   32    Male   29           High  160    7.8                         No   
127   37  Female   34           High  180    8.5                         No   

    Smoking     Diet Exercise Diagnosis  
0        No  Healthy  Regular        No  
1       Yes     Poor       No       Yes  
2       Yes     Poor       No       Yes  
3       Yes     Poor       No       Yes  
4        No  Healthy  Regular        No  
..      ...      ...      ...       ...  
123     Yes     Poor       No       Yes  
124     Yes     Poor       No       Yes  
125     Yes     Poor       No       Yes  
126     Yes     Poor       No       Yes  
127     Yes     Poor       No       Yes  

[128 rows x 11 columns]

**Note**: No missing values in this dataset

##Categorical values

In [55]:
x=pd.get_dummies(x).drop(['Gender_Male','Blood Pressure_High','Family History of Diabetes_No','Smoking_No','Diet_Poor','Exercise_No'], axis=1)
x.head(5)

Age  BMI  FBS  HbA1c  Gender_Female  Blood Pressure_Low  \
0   45   25  100    5.7              0                   0   
1   55   30  120    6.4              1                   0   
2   65   35  140    7.1              0                   0   
3   75   40  160    7.8              1                   0   
4   40   20   80    5.0              0                   0   

   Blood Pressure_Normal  Family History of Diabetes_Yes  Smoking_Yes  \
0                      1                               0            0   
1                      0                               1            1   
2                      0                               1            1   
3                      0                               1            1   
4                      1                               0            0   

   Diet_Healthy  Exercise_Regular  
0             1                 1  
1             0                 0  
2             0                 0  
3             0                 0  
4             1                 1

In [56]:
x.shape

(128, 11)

In [62]:
# Normalization of data
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x = sc.fit_transform(x)

In [63]:
x

array([[ 0.17757568, -0.69418375, -1.02318189, ..., -1.26974206,
         1.26974206,  1.26974206],
       [ 0.77572533, -0.35913277, -0.69576368, ...,  0.78756153,
        -0.78756153, -0.78756153],
       [ 1.37387498, -0.02408179, -0.36834548, ...,  0.78756153,
        -0.78756153, -0.78756153],
       ...,
       [-0.89909369, -0.76119395, -0.36834548, ...,  0.78756153,
        -0.78756153, -0.78756153],
       [-0.60001887, -0.42614297, -0.04092728, ...,  0.78756153,
        -0.78756153, -0.78756153],
       [-0.30094404, -0.09109199,  0.28649093, ...,  0.78756153,
        -0.78756153, -0.78756153]])

In [58]:
y.head(5)

0     No
1    Yes
2    Yes
3    Yes
4     No
Name: Diagnosis, dtype: object

In [59]:
Diagnosis= {'No':0, 'Yes':1}
y = y.apply(lambda x: Diagnosis[x])
y.head()

0    0
1    1
2    1
3    1
4    0
Name: Diagnosis, dtype: int64

In [60]:
y

0      0
1      1
2      1
3      1
4      0
      ..
123    1
124    1
125    1
126    1
127    1
Name: Diagnosis, Length: 128, dtype: int64

#Spliting Dataset

In [68]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

In [71]:
x_train.shape

(102, 11)

In [72]:
x_test.shape

(26, 11)

In [73]:
y_train

91     0
16     0
90     0
118    0
13     1
      ..
67     1
64     0
117    0
47     0
44     0
Name: Diagnosis, Length: 102, dtype: int64

In [76]:
y_test

40     1
24     0
86     0
51     0
8      0
108    0
127    1
96     1
73     0
7      0
60     0
10     0
89     0
94     1
30     0
33     0
2      1
59     0
50     0
26     0
22     0
101    0
48     0
68     1
121    0
122    1
Name: Diagnosis, dtype: int64

#Logistic Regression Model

In [77]:
from sklearn.linear_model import LogisticRegression
model_Logistic = LogisticRegression()
model_Logistic.fit(x_train, y_train)

LogisticRegression()

##Prediction

In [78]:
y_pred_Logistic = model_Logistic.predict(x_test)
print(y_pred_Logistic)

[1 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 1]


In [79]:
print(y_test)

40     1
24     0
86     0
51     0
8      0
108    0
127    1
96     1
73     0
7      0
60     0
10     0
89     0
94     1
30     0
33     0
2      1
59     0
50     0
26     0
22     0
101    0
48     0
68     1
121    0
122    1
Name: Diagnosis, dtype: int64


##Evaluation

In [80]:
from sklearn.metrics import accuracy_score, confusion_matrix
print(accuracy_score(y_test, y_pred_Logistic)*100)
print(confusion_matrix(y_test, y_pred_Logistic))

96.15384615384616
[[19  0]
 [ 1  6]]


#Decision Tree Model

In [81]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier()
classifier.fit(x_train, y_train)

DecisionTreeClassifier()

##Prediction

In [82]:
y_pred_decision_tree = classifier.predict(x_test)
print(y_pred_decision_tree)

[1 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0 1]


In [83]:
y_test

40     1
24     0
86     0
51     0
8      0
108    0
127    1
96     1
73     0
7      0
60     0
10     0
89     0
94     1
30     0
33     0
2      1
59     0
50     0
26     0
22     0
101    0
48     0
68     1
121    0
122    1
Name: Diagnosis, dtype: int64

##Evaluation

In [84]:
print(accuracy_score(y_test, y_pred_decision_tree)*100)
print(confusion_matrix(y_test, y_pred_decision_tree))

100.0
[[19  0]
 [ 0  7]]


#Random Forest

In [85]:
from sklearn.ensemble import RandomForestClassifier
classifier2 = RandomForestClassifier(n_estimators=100, random_state=1)
classifier2.fit(x_train, y_train)

RandomForestClassifier(random_state=1)

##Prediction

In [87]:
y_pred_random_forest = classifier2.predict(x_test)
print(y_pred_random_forest)

[1 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 1]


In [88]:
print(y_test.values)

[1 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 1 0 1]


##Evaluation

In [89]:
print(accuracy_score(y_test, y_pred_random_forest)*100)
print(confusion_matrix(y_test, y_pred_random_forest))

96.15384615384616
[[19  0]
 [ 1  6]]


#XGBoost

In [90]:
from xgboost import XGBClassifier
classifier3 = XGBClassifier()
classifier3.fit(x_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=None, n_jobs=None,
              num_parallel_tree=None, random_state=None, ...)

##Prediction

In [91]:
y_pred_xgboost = classifier3.predict(x_test)
print(y_pred_xgboost)

[1 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 1 0 1]


In [92]:
print(y_test)

40     1
24     0
86     0
51     0
8      0
108    0
127    1
96     1
73     0
7      0
60     0
10     0
89     0
94     1
30     0
33     0
2      1
59     0
50     0
26     0
22     0
101    0
48     0
68     1
121    0
122    1
Name: Diagnosis, dtype: int64


##Evaluation

In [93]:
print(accuracy_score(y_test, y_pred_xgboost)*100)
print(confusion_matrix(y_test, y_pred_xgboost))

96.15384615384616
[[19  0]
 [ 1  6]]


#Conclusion

After doing thes 4 models for Diabetes dataset we can see that:





*   Logistic Regression Model has an accuracy: 96.15384615384616


*   Decision Tree Model has an accuracy: 100.0


*   Random Forest Model has an accuracy: 96.15384615384616


*  XGBoost Model has an accuracy: 96.15384615384616



All model are very good to predict if the person has diabetes or not, because they have high accuracy. The best one is **Decision tree** model.